resizable header in editor

tophf opened the pull request:
This one seems trivial but I'm uploading it as a PR for testing/reviewing in case I'm wrong.
Only the editor is affected because the style manager doesn't seem to benefit from this feature.

Fixes #398.

eight04 reviewed (changes requested):
Tested with Firefox Dev 96.0b10. The header width jumps around when dragging. Probably relate to the `lastW` issue.

tophf commented on edit/header-resizer.js:
```diff
@@ -0,0 +1,50 @@
+  const PREF_ID = 'editor.headerWidth';
+  const el = $('#header-resizer');
+  let lastW, lastX;
+  prefs.subscribe(PREF_ID, (key, val) => {
```
updateHeaderWidth is in `base` because it must run early before the page is shown so I can't move it into header-resizer which runs after everything was shown.

I also don't want to embed more code from header-resizer into base because "base" means the basic necessities e.g. there's no need to process codemirror widgets in base because there's no codemirror yet.

tophf commented on edit/header-resizer.js:
```diff
@@ -0,0 +1,50 @@
+  function setWidth(w) {
+    const delta = (w = editor.updateHeaderWidth(w)) - lastW;
+    if (!delta) return;
+    for (const el of $$('.CodeMirror-linewidget[style*="width:"]')) {
```
CodeMirror calculates this inline style attribute internally each time it redraws the view depending on what else is shown e.g. gutters. There are too many dependencies so I don't think we can use pure css for this.

tophf commented on edit/edit.css:
```diff
@@ -1,12 +1,25 @@
 :root {
   --fixed-padding: unset;
+  --header-width: 280px;
```
The editor should have its own header width so probably not.

tophf commented:
> Firefox Dev 96.0b10. The header width jumps around

Confirmed. FF95 is fine so maybe this is a bug in FF96, gonna investigate...

tophf commented on edit/header-resizer.js:
```diff
@@ -0,0 +1,50 @@
+  const el = $('#header-resizer');
+  let lastW, lastX;
+  prefs.subscribe(PREF_ID, (key, val) => {
+    setLastWidth();
```
Because lastW is used by `setWidth` while `lastX` is not. Maybe a better name would be `currentW` or `actualW`.

tophf commented:
Nah, my code was broken, most likely after adding prefs.subscribe. Fixed now.

tophf commented:
Now the header is resizable also in the manager and the installer.

tophf commented on js/dom-on-load.js:
```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+/** DOM housekeeping after a page finished loading */
+
```
This file is extracted from dom.js and runs after `load` event. The only changes are: 1) `document` is replaced with `window` for some events so that all listeners are uniform, and 2) listeners on `load` event were "unpacked" as well as one debounced call.

tophf commented:
This tooltip is shown when hovering the resizer:

> Holding Shift key while dragging will set the new size in all UI pages

Is it sufficiently clear?
Maybe a direct action verb is better e.g. "Hold Shift key while dragging to set the new size in all UI pages"?

tophf commented:
Maybe it should be reversed so the same width is set by default everywhere, while `Shift` limits it to the current page type?

tophf commented:
Yeah, reversed Shift's behavior because the same width looks much smoother when going between the manager and editor.

The new tooltip explains the concept better, hopefully:
>Hold Shift to resize only in this type of UI, i.e. editor, manager, installer

eight04 commented:
So there are two variables (global, local) controlling the width? Can users unset the local one?

narcolepticinsomniac commented:
My bad, I obviously don't monitor notifications as well as I should. I tend to loosely follow along on the Discord feed more reliably, so if you ever really wanna get my attention, just ping me early in a comment, and it'll show up there.

I get the appeal of providing this feature, so minor adjustments can be made, in terms of awkward alignment issues caused by longer text strings in foreign languages. 

If that's the case, probably makes more sense to allow for them to be set independently, on the different page types.

Don't see how anyone would ever wanna set it to 50%, but you I guess we gotta pick some fairly arbitrary limit. 

Depending how wide we allow headers to be set, might wanna think about setting some max-widths on those flex inputs/selects. At some point they start to look ridiculous.

tophf commented:
> probably makes more sense to allow for them to be set independently, on the different page types

That's what I implemented initially. Then I changed my mind after a while because going from manager to editor and back looked bumpy due to the change of the header width. It's still possible to set different widths per page by holding Shift.

> setting some max-widths on those flex inputs/selects

Well... how about only if someone who actually uses such a wide setting mentions the issue?

narcolepticinsomniac commented:
> going from manager to editor and back looked bumpy

Decent reason.

>  It's still possible to set different widths per page by holding Shift.

Fair enough.

> if someone who actually uses such a wide setting mentions the issue

Aha, if no one notices or cares, is it really an issue? Hard to argue with that logic. =)

tophf commented:
I see `Stylus DeepDark` theme doesn't work well with this so @RaitaroH could you rename your variable to `--header-width` and publish an update now so that your theme doesn't break in the next Stylus update?

eight04 commented on js/header-resizer.js:
```diff
@@ -0,0 +1,63 @@
+  /** @param {MouseEvent} e */
+  function resize(e) {
+    if (setWidth(offset + e.clientX)) {
+      debounce(save, 250, e.shiftKey);
```
We used to update prefs at real-time. Now it only updates prefs when the mouse stopped. Is this intended?

eight04 replied:
If we don't want to save at real-time, it might be better to save on mouseup. Otherwise, users may get some confusing results if they tap the shift key while dragging.

Or, use the shift key from mousedown event.

eight04 commented on js/header-resizer.js:
```diff
@@ -0,0 +1,63 @@
+  let curW, offset, active;
+  prefs.subscribe(dom.HWprefId, (key, val) => {
+    if (!active && val !== curW) {
+      getCurWidth();
```
It seems that we can remove this line.

When `active`, this line won't be called.

When `!active`, this is only used to decide whether or not to update CodeMirror widgets. The `delta` check at L55 seems also partly overlap with L9.

tophf replied:
Yeah.

eight04 commented on js/header-resizer.js:
```diff
@@ -0,0 +1,63 @@
+    const delta = (w = dom.setHWProp(w)) - curW;
+    if (delta) {
+      curW = w;
+      for (const el of $$('.CodeMirror-linewidget[style*="width:"]')) {
```
In the future, we may want to switch to `ResizeObserver` (probably in `codemirror-factory.js`) so it will always work not only when resizing the sidebar.

tophf replied:
In which case it'd be better? Also ResizeObserver is not implemented in our minimum supported browsers.

eight04 replied:
> In which case it'd be better?

Pros after moving this CodeMirror code:

* The logic is decoupled from `header-resizer.js`.
* It is future-proof so that it can handle all kinds of editor resize.

It will be even better if this can be implemented in CodeMirror. Though I guess the dev would suggest us to refresh manually.

> ResizeObserver is not implemented in our minimum supported browsers.

Yeah that's the problem if we want to implement it now.

tophf replied:
I've tried extracting it and use MutationObserver on document.documentElement but I didn't like this solution. It's bigger, slower, more complicated compared to this trivial two-liner.

eight04 reviewed (commented):
LGTM. Tested with FF Dev 96.0b7

tophf commented on js/dom.js:
```diff
@@ -459,42 +462,33 @@ async function waitForSheet({
+      HW,
+      HWprefId,
+      setHWProp(width) {
+        width = Math.round(Math.max(200, Math.min(innerWidth / 2, Number(width) || 0)));
```
We can't control it via CSS because we use this variable for stuff other than `width` e.g. padding or flex-base so we would need `min` and `max` in CSS but these aren't implemented in our minimum supported browsers.

tophf commented on js/dom.js:
```diff
@@ -459,42 +462,33 @@ async function waitForSheet({
+        return width;
+      },
+    });
+    prefs.ready.then(() => dom.setHWProp(prefs.get(HWprefId)));
```
The subscribed listener uses a function from header.resizer.js so I don't see why it'd be better to move it. I want to extract as much as possible from dom.js

tophf commented:
Thanks, I've simplified the code accordingly.
